### PR TITLE
refactor build configuration + don't configure Android projects locally

### PR DESCRIPTION
**Pull request by dsokal**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

This PR simplifies the steps needed to configure a project to be built with EAS Build.
Currently, we still require the Android project to be configured "locally". That is EAS CLI adds the `eas-build.gradle` file to the project and modifies the `build.gradle`. This way we can sign the project with the keystore provided to EAS Build. We're moving the script injection to EAS workers. 

The PR also simplifies the configuration around `expo-updates`. Currently, the configuration consists of two steps:
- setting updates URL, `fallbackToCacheTimeout`, etc.
- synchronizing runtime/sdk version

It turns out the function we use for completing the first step is also covering the second step. Instead of having two steps we can always call the more generic function.

In the future, we might want to get rid of the `eas build:configure` command entirely. All the steps are covered by `eas build` as well. Removing the command can be as simple as removing the `build/configure.ts` file.

# How

- Remove `eas-build.gradle` and `apply from: "./eas-build.gradle"` from `build.gradle` when running `build`/`build:configure`.
- Remove `--skip-project-configuration` flag from `eas build`. It seems we don't need it anymore as we don't require commits (it depends on the user configuration).
- Always configure `expo-updates` (if installed in the project) when running `build`/`build:configure`.

# Test Plan

Tested manually with the following project types:
- New project that has `expo-updates` installed.
- New project that doesn't have `expo-updates` installed.
- Old project that has `eas-build.gradle` file. 

I tested configuring them with both `eas build:configure` and `eas build`.

### CODE REVIEW

**Review by jkhales — APPROVED** [review]
This is a great clean up.

I'm curious about how we are thinking to handle the potential native vs app.json conflicts. But assuming there is a good reason, I have no other questions!

File: packages/eas-cli/src/build/build.ts [hunk 1]
```diff
@@ -60,14 +60,13 @@ export async function prepareBuildRequestForPlatformAsync<
       trackingCtx: ctx.trackingCtx,
     }
   );
-  if (!ctx.skipProjectConfiguration) {
```
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/build/build.ts:63]: No, I decided to remove the `--skip-project-configuration` flag from `eas build` (see the PR description).

File: packages/eas-cli/src/build/android/UpdatesModule.ts [hunk 2]
```diff
@@ -1,63 +1,29 @@
...
-}
-
 export async function syncUpdatesConfigurationAsync(
   projectDir: string,
   exp: ExpoConfig
 ): Promise<void> {
   ensureValidVersions(exp);
   const accountName = getProjectAccountName(exp, await ensureLoggedInAsync());
-  try {
-    await ensureUpdatesConfiguredAsync(projectDir);
```
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/build/android/UpdatesModule.ts:30]: Previously, you could run `eas build --skip-project-configuration` so the updates wouldn't be configured automatically. Now, we don't have the flag so updates are always configured (if expo-updates is installed).

File: packages/eas-cli/src/build/android/UpdatesModule.ts [hunk 3]
```diff
@@ -1,63 +1,29 @@
...
 
   const androidManifestPath = await AndroidConfig.Paths.getAndroidManifestAsync(projectDir);
-  let androidManifest = await getAndroidManifestAsync(projectDir);
-
-  if (!AndroidConfig.Updates.areVersionsSynced(exp, androidManifest)) {
-    androidManifest = AndroidConfig.Updates.setVersionsConfig(exp, androidManifest);
-    await AndroidConfig.Manifest.writeAndroidManifestAsync(androidManifestPath, androidManifest);
-  }
-
-  if (!AndroidConfig.Updates.isMainApplicationMetaDataSynced(exp, androidManifest, accountName)) {
```
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/build/android/UpdatesModule.ts:46]: We always call `AndroidConfig.Updates.setUpdatesConfig` (no matter what) so the metadata are always synced.

File: packages/eas-cli/src/build/ios/UpdatesModule.ts [hunk 4]
```diff
@@ -1,60 +1,21 @@
...
-}
-
 export async function syncUpdatesConfigurationAsync(
   projectDir: string,
   exp: ExpoConfig
 ): Promise<void> {
   ensureValidVersions(exp);
   const accountName = getProjectAccountName(exp, await ensureLoggedInAsync());
-  try {
-    await ensureUpdatesConfiguredAsync(projectDir);
```
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/build/ios/UpdatesModule.ts:30]: See the thread for the android counterpart.

File: packages/eas-cli/src/build/ios/UpdatesModule.ts [hunk 5]
```diff
@@ -1,60 +1,21 @@
...
-    throw error;
-  }
-
-  let expoPlist = await readExpoPlistAsync(projectDir);
-  if (!IOSConfig.Updates.isPlistVersionConfigurationSynced(exp, expoPlist)) {
-    expoPlist = IOSConfig.Updates.setVersionsConfig(exp, expoPlist);
-    await writeExpoPlistAsync(projectDir, expoPlist);
-  }
-
-  if (!IOSConfig.Updates.isPlistConfigurationSynced(exp, expoPlist, accountName)) {
```
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/build/ios/UpdatesModule.ts:44]: See the thread for the android counterpart.

File: packages/eas-cli/src/build/android/syncProjectConfiguration.ts [hunk 6]
```diff
@@ -73,3 +45,22 @@ function resolveVersionBumpStrategy(autoIncrement: AndroidVersionAutoIncrement):
...
+// TODO: remove this after a few months
+export async function cleanUpOldEasBuildGradleScriptAsync(projectDir: string): Promise<void> {
+  const easBuildGradlePath = path.join(projectDir, 'android', 'app', 'eas-build.gradle');
+  if (await fs.pathExists(easBuildGradlePath)) {
+    await fs.remove(easBuildGradlePath);
+
+    const buildGradlePath = AndroidConfig.Paths.getAppBuildGradleFilePath(projectDir);
+    const buildGradleContents = await fs.readFile(buildGradlePath, 'utf-8');
+    const APPLY_EAS_BUILD_GRADLE_LINE = 'apply from: "./eas-build.gradle"';
+    const buildGradleContentsWithoutApply = buildGradleContents.replace(
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/android/syncProjectConfiguration.ts:58]: I think that apply can also use single quotes

File: packages/eas-cli/src/build/android/syncProjectConfiguration.ts [hunk 7]
```diff
@@ -73,3 +45,22 @@ function resolveVersionBumpStrategy(autoIncrement: AndroidVersionAutoIncrement):
...
+      `${APPLY_EAS_BUILD_GRADLE_LINE}\n`,
+      ''
+    );
+    if (buildGradleContentsWithoutApply !== buildGradleContents) {
+      await fs.writeFile(buildGradlePath, buildGradleContentsWithoutApply);
+    }
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/android/syncProjectConfiguration.ts:64]: add some log(maybe warning) that we are removing that file

File: packages/eas-cli/src/build/configure.ts [hunk 8]
```diff
@@ -138,58 +88,19 @@ const EAS_JSON_BARE_DEFAULT: EasJson = {
...
+async function createEasJsonAsync(projectDir: string): Promise<void> {
+  const easJsonPath = EasJsonReader.formatEasJsonPath(projectDir);
 
+  const hasAndroidNativeProject =
+    (await resolveWorkflowAsync(projectDir, Platform.ANDROID)) === Workflow.GENERIC;
+  const hasIosNativeProject =
+    (await resolveWorkflowAsync(projectDir, Platform.IOS)) === Workflow.GENERIC;
   const easJson =
-    ctx.hasAndroidNativeProject && ctx.hasIosNativeProject
+    hasAndroidNativeProject || hasIosNativeProject
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/configure.ts:99]: I think previous behavior was better for users, if one of the platforms is bare and the other is managed then bare config for both is more readable.

**Review by wkozyra95 — APPROVED** [review]

**Review by wkozyra95 — APPROVED** [review]

**Review by brentvatne — APPROVED** [review]

**Pull request merged**